File: splitsy-frontend/src/domain/expense/value-objects/Split.ts
import { ValueObject } from "../../shared/value-objects/ValueObject";

export interface SplitProps {
  participantId: number | string;
  amount: number;
}

export class Split extends ValueObject<SplitProps> {
  private constructor(props: SplitProps) {
    super(props);
  }

  public static create(participantId: number | string, amount: number): Split {
    if (participantId === null || participantId === undefined || participantId === '') {
      throw new Error("ParticipantId missing in split");
    }

    if (amount === null || amount === undefined) {
      throw new Error("The split needs an amount");
    }

    if (amount <= 0) {
      throw new Error("Split amount must be greater than 0");
    }

    return new Split({ participantId, amount });
  }

  public get participantId(): number | string {
    return this.props.participantId;
  }

  public get amount(): number {
    return this.props.amount;
  }

  public toJSON() {
    return {
      participantId: this.participantId,
      amount: this.amount,
    };
  }
}

File: splitsy-backend/src/domain/expenses/value_objects/split.py
from dataclasses import dataclass

from src.domain.common.value_objects import Amount, Id, ValueObject


@dataclass(frozen=True)
class SplitProps:
    participant_id: Id
    amount: Amount


class Split(ValueObject[SplitProps]):
    def __init__(self, props: SplitProps) -> None:
        super().__init__(props)
        self._participant_id = props.participant_id
        self._amount = props.amount

    @property
    def participant_id(self) -> Id:
        return self._participant_id

    @property
    def amount(self) -> Amount:
        return self._amount

    @staticmethod
    def create(participant_id: Id, amount: Amount) -> "Split":
        return Split(SplitProps(participant_id=participant_id, amount=amount))
